stateful operations using callable classes

henrychi2 opened the pull request:
Stateful operations for filter(), transform(), multi_transform() via instances of callable classes

By allowing filter / transform function operators to accept instances of callable classes as input, state information such as user-defined parameters can be stored during class initialization and utilized when the class is called.

Miscellaneous fixes:
-Initialize function pointers to NULL in filter / transform custom operators
-Decrement function pointer reference count if necessary in operator destructor
-Additional error checking in custom operators
-Do not share `_hwcount` global variable among test functions

ddebrunner commented:
I looked quickly, but I didn't see anything explaining what the api is for the developer.

Is there an overview of what is being implemented here.

ddebrunner commented:
Can one of the admins verify this patch?

wmarshall484 commented:
@henrychi2, maybe add a section in the spldoc like in pull request #283

wmarshall484 commented on com.ibm.streamsx.topology/com.ibm.streamsx.topology.functional.python/PyFunctionMultiTransform/PyFunctionMultiTransform_cpp.cgt:
```diff
@@ -130,13 +130,30 @@ void MY_OPERATOR::process(Tuple const & tuple, uint32_t port)
+  
+  // first argument is application function or serialized callable class
+  <% if ($serializedClass) { %>
+    PyTuple_SetItem(funcArg, 0, Py_BuildValue("s", <%=$serializedClass%>));
```
Deserializing the callable class should be done once in the consturctor, not every time in the process method.

ddebrunner replied:
Is this serializing a *class* or an *instance of a class* ?

I wouldn't expect we need to serialize a *class*.

wmarshall484 replied:
Instance of a class.

henrychi2 replied:
1) Fixed so that deserializing the instance of the callable class is done once in the constructor.
2) Changed comments to state that we are serializing an instance of a class.

henrychi2 commented on test/python/topology/test_functions.py:
```diff
@@ -52,56 +52,112 @@ def string_to_int_except68(t):
+global EXPECTED_STRINGS_LENGTH_FILTER
+EXPECTED_STRINGS_LENGTH_FILTER = ["goodbye", "farewell"]
+
+global _hwcount_length_filter
```
@wmarshall484 
Currently I define a global counter for each new testcase.  To avoid this, do you think it makes sense to support parameters for sink() as well (maybe source() too) ?  Thanks.

ddebrunner replied:
No, the counters are a work-around for the fact that the testing infrastructure that the Java Application API supports is not yet supported by Python.

henrychi2 commented:
Added some documentation and usage example to spldoc.

ddebrunner commented on com.ibm.streamsx.topology/com.ibm.streamsx.topology.functional.python/PyFunctionMultiTransform/PyFunctionMultiTransform.xml:
```diff
@@ -49,13 +49,22 @@
         <cardinality>1</cardinality>
       </parameter>
+      <parameter>
+        <name>serializedClass</name>
```
So it's not a 'serializedClass' so can the name reflect what it represents.

henrychi2 replied:
Change parameter and variable names to 'serializedCallableInstance'.

ddebrunner commented on com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/runtime.py:
```diff
@@ -22,11 +23,24 @@ def _pickleReturn(v):
+# base64 encoded, return a function that
+# depickles the input and then calls F
+def depickleInputForCallable(serializedCallable) :
+    callableObject = pickle.loads(base64.b64decode(serializedCallable))
```
Should this use the function declared above? `depickleCallable`

henrychi2 replied:
Fixed in latest commit.

ddebrunner commented on com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/topology.py:
```diff
@@ -49,9 +49,14 @@ def sink(self, func):
+        Filters tuples from a stream using the supplied callable 'func'.
+        The callable input must be either 
+        1) the name of a top-level function defined using 'def' that takes a single parameter for the tuple, or
+        2) an instance of a top-level callable class that implements the method '__call__(self, tuple)'.  Using a
```
What's a "top-level" function or class?

I don't see a standard definition in my python book.

I found this: https://docs.python.org/3.5/reference/toplevel_components.html
but it never defines what top-level means.

wmarshall484 replied:
> the name of a top-level function defined using 'def'
How about "the name of a function defined in a module"?

henrychi2 replied:
Changed the comment to read "functions defined at the top level of a module".  This is the same terminology used in Python doc here:
https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled

All I want to say here is that we don't support nested functions, since an inner function is local the outer function and thus not visible by name to PythonC.

ddebrunner commented on com.ibm.streamsx.topology/opt/python/packages/streamsx/topology/topology.py:
```diff
@@ -62,10 +67,15 @@ def filter(self, func):
-        Transforms each tuple from this stream into 0 or 1 tuples using the supplied function.
+        Transforms each tuple from this stream into 0 or 1 tuples using the supplied callable 'func'.
+        The callable input must be either 
+        1) the name of a top-level function defined using 'def' that takes a single parameter for the tuple, or
```
Is there another way of defining functions that doesn't use 'def'?

henrychi2 replied:
I guess lambda functions can also be 'defined', which we don't support.
Removed the comment 'using def', since we already mention that we don't support lambdas in the spldoc.

ddebrunner commented:
At some point tests that have functions with mutable state are needed, can be a follow on pull request.

wmarshall484 commented:
I'd be curious to see whether you could pass a lambda to `filter` or `transform`. As far as I know, lambdas are callable objects, so they might work with the current code.

henrychi2 commented:
As a sanity check, I added a testcase with mutable state:

```
class IncMaxSplitWords:
   def __init__(self, maxsplit):
      self.maxsplit = maxsplit  
   def __call__(self, tuple):
      words = tuple.split(None, self.maxsplit)
      # test mutable state, increase maxsplit by 1 for next tuple
      self.maxsplit += 1
      return words
```

henrychi2 commented:
Unfortunately it doesn't seem like lambdas work with `filter` or `transform` in the current implementation.  This is because lambdas are anonymous, so in PythonC we don't have a reference to a function name.
`Fatal error: function not callable: <lambda>`

Another way is to somehow serialize the lambda.  But pickle doesn't support this.  Doing some search on the web, lambdas might be serialized with marshal, but it might have limitations.  Or lambas can be serialized with third-party libraries like dill.  This will require more investigation in the future.

ddebrunner commented on com.ibm.streamsx.topology/com.ibm.streamsx.topology.functional.python/PyFunctionTransform/PyFunctionTransform_cpp.cgt:
```diff
@@ -62,14 +64,27 @@ MY_OPERATOR::MY_OPERATOR()
+    <% if ($serializedCallableInstance) { %>
+      depickleInput = streamsx::topology::Splpy::loadFunction("streamsx.topology.runtime", "depickleInputForCallableInstance");
+      PyTuple_SetItem(funcArg, 0, Py_BuildValue("s", <%=$serializedCallableInstance%>));
+      Py_DECREF(appFunction);
```
I found trying to understand the use of appFunction hard, it used to be the function that was going to be called (or wrapped by another function), but what is appFunction in the case it's a callable class? What is functionName set to when the function is loaded at line 62?

ddebrunner commented:
I think the code would become clearer if the operator parameter names reflected what they are, rather than what they used to be.

`functionName --> name` ??? Maybe something better here. `pyName` ??

`functionModule --> module`

`serializedCallableInstance --> `callable` - Just easier to talk about than something with 7 syllables .

The current code is hard to understand because 'functionName' represents a function name or a class name, but a reader naturally assumes it represents only a function name.

ddebrunner commented:
Will there be a follow pull request for Sink and Source?

henrychi2 commented:
@ddebrunner 
1) Sure, I will change the operator parameter names.
2) Just want to confirm...we want Sink and Source to accept instances of callable classes correct?  Let me know if I misunderstood, otherwise I can start working on that and include in this pull request.

ddebrunner commented:
I think Sink/Source can be a follow on pull request, let's get this one merged first.

henrychi2 commented:
Changed the operator parameter names with latest commit:

`functionName --> pyName`
`functionModule --> pyModule`
`serializedCallableInstance --> pyCallable`
